lib/fastly_nsq/testing.rb

```diff
@@ -0,0 +1,239 @@
+# frozen_string_literal: true
+
+module FastlyNsq
+  class Testing
+    class << self
+      attr_accessor :__test_mode
+
+      def __set_test_mode(mode)
+        if block_given?
+          current_mode = __test_mode
+          begin
+            self.__test_mode = mode
+            yield
+          ensure
+            self.__test_mode = current_mode
+          end
+        else
+          self.__test_mode = mode
+        end
+      end
+
+      def disable!(&block)
+        __set_test_mode(:disable, &block)
+      end
+
+      def fake!(&block)
+        __set_test_mode(:fake, &block)
+      end
+
+      def inline!(&block)
+        __set_test_mode(:inline, &block)
+      end
+
+      def enabled?
+        __test_mode != :disable
+      end
+
+      def disabled?
+        __test_mode == :disable
+      end
+
+      def fake?
+        __test_mode == :fake
+      end
+
+      def inline?
+        __test_mode == :inline
+      end
+
+      def reset!
+        return unless enabled?
+        FastlyNsq::Messages.messages.clear
+      end
+    end
+  end
+
+  module Messages
+    def self.messages
+      @messages ||= Hash.new { |h, k| h[k] = [] }
```

Comment:
pedantic: `Hash.new([])`

Reply:
I like it.

```
= Hash.new

(from ruby core)
------------------------------------------------------------------------------
  Hash.new                          -> new_hash
  Hash.new(obj)                     -> new_hash
  Hash.new {|hash, key| block }     -> new_hash

------------------------------------------------------------------------------

Returns a new, empty hash. If this hash is subsequently accessed by a key that
doesn't correspond to a hash entry, the value returned depends on the style of
new used to create the hash. In the first form, the access returns nil. If
obj is specified, this single object will be used for all
default values. If a block is specified, it will
be called with the hash object and the key, and should return the default
value. It is the block's responsibility to store the value in the hash if
required.

  h = Hash.new("Go Fish")
  h["a"] = 100
  h["b"] = 200
  h["a"]           #=> 100
  h["c"]           #=> "Go Fish"
  # The following alters the single default object
  h["c"].upcase!   #=> "GO FISH"
  h["d"]           #=> "GO FISH"
  h.keys           #=> ["a", "b"]

  # While this creates a new default object each time
  h = Hash.new { |hash, key| hash[key] = "Go Fish: #{key}" }
  h["c"]           #=> "Go Fish: c"
  h["c"].upcase!   #=> "GO FISH: C"
  h["d"]           #=> "Go Fish: d"
  h.keys           #=> ["c", "d"]
```

Reply:
The documentation here is poor but `Hash.new([])` has a different behavior.

>  If obj is specified, this single object will be used for all default values.

```ruby
[1] pry(main)> h = Hash.new([])
=> {}
[2] pry(main)> h['foo'] << 'bar'
=> ["bar"]
[3] pry(main)> h
=> {}
```

> If a block is specified, it will be called with the hash object and the key, and should return the default value. It is the block's responsibility to store the value in the hash if
required.

```ruby
[4] pry(main)> h = Hash.new { |h,k| h[k] = [] }
=> {}
[5] pry(main)> h['foo'] << 'bar'
=> ["bar"]
[6] pry(main)> h
=> {"foo"=>["bar"]}
```

Reply:
umm, yea that is striking.  Carry on.